src/main/java/org/openjdk/jextract/impl/Utils.java

```diff
@@ -100,56 +104,73 @@ static String quote(char ch) {
         return switch (scoped.kind()) {
             case STRUCT, UNION -> true;
             default -> false;
         };
     }
 
-    static boolean isPointerType(Type type) {
-        if (type instanceof Delegated delegated) {
-            return delegated.kind() == Delegated.Kind.POINTER;
-        } else {
-            return false;
-        }
+    static boolean isEnum(Declaration.Scoped scoped) {
+        return switch (scoped.kind()) {
+            case ENUM -> true;
+            default -> false;
+        };
+    }
+
+    static boolean isArray(Type type) {
+        return switch (type) {
+            case Type.Array _ -> true;
+            case Type.Delegated delegated when delegated.kind() == Delegated.Kind.TYPEDEF ->
+                    isArray(delegated.type());
+            default -> false;
+        };
+    }
+
+    static boolean isEnum(Type type) {
+        return switch (type) {
+            case Type.Declared declared -> isEnum(declared.tree());
+            case Type.Delegated delegated when delegated.kind() == Delegated.Kind.TYPEDEF ->
+                    isEnum(delegated.type());
+            default -> false;
+        };
+    }
+
+    static boolean isStructOrUnion(Type type) {
+        return switch (type) {
+            case Type.Declared declared -> isStructOrUnion(declared.tree());
+            case Type.Delegated delegated when delegated.kind() == Delegated.Kind.TYPEDEF ->
+                isStructOrUnion(delegated.type());
+            default -> false;
+        };
+    }
+
+    static boolean isPointer(Type type) {
+        return switch (type) {
+            case Type.Delegated delegated when delegated.kind() == Delegated.Kind.TYPEDEF ->
+                    isPointer(delegated.type());
+            case Type.Delegated delegated when delegated.kind() == Delegated.Kind.POINTER ->
+                    true;
+            default -> false;
+        };
+    }
+
+    static boolean isPrimitive(Type type) {
+        return switch (type) {
+            case Type.Declared declared when declared.tree().kind() == Declaration.Scoped.Kind.ENUM ->
+                isPrimitive(((Declaration.Constant)declared.tree().members().get(0)).type());
+            case Type.Delegated delegated -> isPrimitive(delegated.type());
+            case Type.Primitive _ -> true;
+            default -> false;
+        };
     }
 
     static Function getAsFunctionPointer(Type type) {
-        if (type instanceof Function function) {
-            /*
-             * // pointer to function declared as function like this
-             *
-             * typedef void CB(int);
-             * void func(CB cb);
-             */
-            return function;
-        } else if (isPointerType(type)) {
-            return getAsFunctionPointer(((Delegated)type).type());
-        } else {
-            return null;
-        }
+        return switch (type) {
+            case Type.Delegated delegated -> getAsFunctionPointer(delegated.type());
```

Comment:
Not sure here - this function is meant to follow pointers until a function type is found. This method seems to do that?

Reply:
I was under the impression that this should only find function types for these 2 cases:

   void (*cb)(void)
   void cb(void)

In which case we want to generate a functional interface. The old code also doesn't seem to follow typedefs for instance, but the new code does. This seems like a behavior change.

Reply:
Yes, it's a behavior change, but when I wrote the new code, it seemed to me that the old behavior was missing certain cases where we wanted to generate a functional interface accessor?

Reply:
Yeah... I think so. I don't think we can get into a situation where we try to generate the same interface twice, since for function parameters, the name of the interface is derived from the name of the function, so there's no clash with typedefs possible (and function param names are unique).